### PR TITLE
add subcase to test case setvcpu

### PR DESCRIPTION
**Pull request by brucepatricklee**
check the vcpu order when hot plug/unplug specific vcpu

Signed-off-by: Jin Li <jil@redhat.com>

### CODE REVIEW

**Comment by brucepatricklee** [comment]
[root@rhel7 tp-libvirt]# avocado run  --vt-type libvirt type_specific.io-github-autotest-libvirt.setvcpu
JOB ID     : 2f4036083f028fba4ff8af0412be1103aff95886
JOB LOG    : /root/avocado/job-results/job-2019-03-29T10.24-2f40360/job.log
 (01/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.coldplug.enable: PASS (13.84 s)
 (02/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.coldplug.disable: PASS (19.74 s)
 (03/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.hotplug.enable: PASS (23.20 s)
 (04/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.hotplug.enable_config: PASS (21.33 s)
 (05/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.hotplug.disable: PASS (22.92 s)
 (06/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.hotplug.disable_config: PASS (21.37 s)
 (07/12) type_specific.io-github-autotest-libvirt.setvcpu.positive_test.hotplug.order: PASS (29.37 s)
 (08/12) type_specific.io-github-autotest-libvirt.setvcpu.negative_test.hotplug_duplicate_vcpu.enable: PASS (21.19 s)
 (09/12) type_specific.io-github-autotest-libvirt.setvcpu.negative_test.hotplug_duplicate_vcpu.disable: PASS (21.24 s)
 (10/12) type_specific.io-github-autotest-libvirt.setvcpu.negative_test.hotplug_invalid_vcpu.disable_online_vcpu: PASS (21.23 s)
 (11/12) type_specific.io-github-autotest-libvirt.setvcpu.negative_test.hotplug_invalid_vcpu.enable_noexist_vcpu: PASS (21.46 s)
 (12/12) type_specific.io-github-autotest-libvirt.setvcpu.negative_test.hotplug_invalid_vcpu.disable_offline_vcpu: PASS (21.21 s)
RESULTS    : PASS 12 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 262.29 s

File: libvirt/tests/src/cpu/setvcpu.py [hunk 1]
```diff
@@ -33,20 +34,20 @@ def run(test, params, env):
     start_timeout = int(params.get("start_timeout", "60"))
     check = params.get("check", "")
     err_msg = params.get("err_msg", "")
-    status_error = "yes" == params.get("status_error", "no")
+    status_error = ("yes" == params.get("status_error", "no"))
```
Comment (chloerh) [inline libvirt/tests/src/cpu/setvcpu.py]: Not necessary

File: libvirt/tests/src/cpu/setvcpu.py [hunk 2]
```diff
@@ -33,20 +34,20 @@ def run(test, params, env):
...
 
     vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
     vmxml_backup = vmxml.copy()
 
-    def check_vcpu_status(cpus_online_pre=1):
+    def check_vcpu_status(cpulist, option, cpus_online_pre=1):
```
Comment (chunfuwen) [inline libvirt/tests/src/cpu/setvcpu.py]: rename option to cpu_option ?

File: libvirt/tests/src/cpu/setvcpu.py [hunk 3]
```diff
@@ -55,25 +56,73 @@ def check_vcpu_status(cpus_online_pre=1):
...
-                    if (vmxml.vcpus.vcpu[cpu_id].get('enabled') != "yes"):
-                        test.fail("vcpu status check fail")
-                elif ("disable" in option):
-                    cpu_count -= 1
-                    if (vmxml.vcpus.vcpu[cpu_id].get('enabled') != "no"):
-                        test.fail("vcpu status check fail")
-                else:
-                    test.fail("wrong vcpu status in xml")
+        for cpu_id in cpulist:
+            if "enable" in option:
```
Comment (chunfuwen) [inline libvirt/tests/src/cpu/setvcpu.py]: maybe wrap into one method :validate_xx()

File: libvirt/tests/src/cpu/setvcpu.py [hunk 4]
```diff
@@ -109,18 +158,36 @@ def check_vcpu_status(cpus_online_pre=1):
...
+                check_vcpu_status(cpulist, option)
 
         # start vm
         virsh.start(vm_name, debug=True, ignore_status=False)
         vm.wait_for_login(timeout=start_timeout)
 
-        cpus_online_pre = vm.get_cpu_count()
+        # turn setvcpu_option to an ordered dict
+        if isinstance(setvcpu_option, tuple):
+            d = collections.OrderedDict()
```
Comment (chunfuwen) [inline libvirt/tests/src/cpu/setvcpu.py]: please check whether there is OrderedDict() on python2

File: libvirt/tests/src/cpu/setvcpu.py [hunk 5]
```diff
@@ -55,25 +56,73 @@ def check_vcpu_status(cpus_online_pre=1):
...
+        vmxml_new = vm_xml.VMXML.new_from_dumpxml(vm_name)
+        # get vcpus order dict from previous xml
+        order_pre = get_vcpu_order(vmxml_pre)
+        # get vcpus order dict from updated xml
+        order_new = get_vcpu_order(vmxml_new)
+
+        # calculate the right dict of vcpu order based on the previous one
+        if "enable" in cpu_option:
+            order_expect = order_pre.copy()
+            order_expect[vcpu] = len(order_pre)+1
```
Comment (chloerh) [inline libvirt/tests/src/cpu/setvcpu.py]: I think we need some space besides '+' and '-'

Comment (brucepatricklee) [reply libvirt/tests/src/cpu/setvcpu.py]: updated

**Comment by brucepatricklee** [comment]
@chloerh updated

**Pull request merged**